common/raw_hashtable.h

```diff
@@ -0,0 +1,1295 @@
+#include "llvm/Support/Compiler.h"
+#include "llvm/Support/MathExtras.h"
+
+// A namespace collecting a set of low-level utilities for building hashtable
+// data structures. These should only be used as implementation details of
+// higher-level data-structure APIs.
+//
+// These utilities support hashtables following a *specific* API design pattern,
+// and using Small-Size Optimization, or "SSO", when desired. We expect there to
+// be three layers to any hashtable design:
+//
+// - A *view* type: a read-only view of the hashtable contents. This type should
+//   be a value type and is expected to be passed by-value in APIs. However, it
+//   will have `const`-reference semantics, much like a `std::string_view`. Note
+//   that the *entries* will continue to be mutable, it is only the *table* that
+//   is read-only.
+//
+// - A *base* type: a base class type of the actual hashtable, which allows
+//   almost all mutable operations but erases any specific SSO buffer size.
+//   Because this is a base of the actual hash table, it is designed to be
+//   passed as a non-`const` reference or pointer.
+//
+// - A *table* type: the actual hashtable which derives from the base type and
+//   adds any desired SSO storage buffer. Beyond the physical storage, it also
+//   allows resetting the table to its initial state & allocated size, as well
+//   as copying and moving the table.
+//
+// For complete examples of the API design, see `set.h` for a hashtable-based
+// set data structure, and `map.h` for a hashtable-based map data structure.
+//
+// The hashtable design implemented here has several key invariants and design
+// elements that are essential to all three of the types above and the
+// functionality they provide.
+//
+// - The underlying hashtable uses [open addressing], a power-of-two table size,
+//   and quadratic probing rather than closed addressing and chaining.
+//
+//   [open addressing]: https://en.wikipedia.org/wiki/Open_addressing
+//
+// - Each _slot_ in the table corresponds to a key, a value, and one byte of
+//   metadata. Each _entry_ is a key and value. The key and value for an entry
+//   are stored together.
+//
+// - The allocated storage is organized into an array of metadata bytes followed
+//   by an array of entry storage.
+//
+// - The metadata byte corresponding to each entry marks that entry is either
+//   empty, deleted, or present. When present, a 7-bit tag is also stored using
+//   another 7 bits form the hash of the entry key.
+//
+// - The storage for an entry is an internal type that should not be exposed to
+//   users, and instead only the underlying keys and values.
+//
+// - The hash addressing and probing occurs over *groups* of entries rather than
+//   individual entries. These are not necessarily buckets of colliding entries,
+//   however, as they may be filled by probing with entries that have different
+//   initial hashed positions in the open addressed table.
```

Comment:
```suggestion
// - The hash addressing and probing occurs over *groups* of slots rather than
//   individual entries. When inserting a new entry, it can be added to the group
//   it hashes to as long it is not full, and can even replace a slot with a
//   tombstone indicating a previously deleted entry. Only when the group is full
//   will it look at the next group in the probe sequence. As a result, groups may
//   have entries that are not the first in the entry's probe sequence. Also, when
//   performing a lookup, every group in the probe sequence must be inspected
//   for the lookup key until it is found or the group has an empty slot.
```